fuel/datasets/hdf5.py

```diff
@@ -64,3 +67,127 @@ def get_data(self, state=None, request=None):
                 raise ValueError
         data = [node[request] for node in self.nodes]
         return data
+
+
+@do_not_pickle_attributes('data_sources')
+class H5PYDataset(Dataset):
+    """An h5py-fueled HDF5 dataset.
+
+    This dataset class assumes a particular file layout:
+
+    * Data sources reside in the root group, and their names define the
+      source names.
+    * The dataset is not explicitly split. Instead, splits are defined as
```

Comment:
For now, but to quote myself:

> If we ever run into a case where we want splits that are non-contiguous, we could use something like h5file.attrs['alternative_train'] = '/split/alternative_train' and store lists of indices in /split/alternative_train, /split/alternative_test, etc. The logic is then:

> * MNIST() loads the HDF5 file in its entirety
* MNIST('train') checks the split = h5file.attrs['train'] attribute and loads
    * if it finds an array it slices h5file[source][split[0]:split[1]]
    * if it finds a string, it loads h5file[source][h5file[split]]